### PR TITLE
Refactor `::partition()` and `::span()`

### LINKED ISSUE #123: `Partition` Operation - Awkward to use?
## Background

I am looking to replace usage of Doctrine Collections with this library in a project, and one of the issues I've come across is the usage of `partition`. Its return type is a bit confusing at first glance and replicating the functionality of the Doctrine Collections method does not seems straightforward for a new user.

## Doctrine Collection

The return type of the [Doctrine Collection Partition](https://github.com/doctrine/collections/blob/1.6.x/lib/Doctrine/Common/Collections/Collection.php#L263) method is very straightforward. It returns an array shape with _exactly two elements_ consisting of the same collection objects with the same  type for the keys and values. Because of this, usage with the PHP [list](https://www.php.net/manual/en/function.list.php) language construct is possible and very convenient:

```PHP
$data = [
    ['id' => 'ABC', 'name' => 'Bob'],
    ['id' => 'BAC', 'name' => 'John'],
    ['id' => 'ABC', 'name' => 'Mike'],
];

[$matches, $nonMatches] = (new ArrayCollection($data))
    ->partition(static fn (int $key, array $entity): bool => $entity['id'] === 'ABC');
```

## Loophp Collection

The return type of the [partition operation](https://github.com/loophp/collection/blob/master/src/Contract/Operation/Partitionable.php#L23) is a collection of lists of arrays of an array shape where the key and value of the elements is included (the nesting is real!).

The first issue is that this operation will always return exactly two elements, but this is not immediately visible from the return type and I'm not sure how this can be fixed because we cannot specify collection shape in the same way as array shape.

The other main issue is the combination of operations needed to achieve the same behaviour as Doctrine Collections: _I just want to be able to get the two separate collections with the values inside them, nothing more._ This is what I've come up with:

```PHP
$partitioned = Collection::fromIterable($data)
    ->partition(static fn(array $entity): bool => $entity['id'] === 'ABC');

$matches = $partitioned->first()->pluck('*.1')->unwrap();
$nonMatches = $partitioned->last()->pluck('*.1')->unwrap();
```

## Possible improvements

A) don't return the keys and the values separately. Rather, they should be together just like in the original collection, before the partition
B) can we reduce the nesting so the `unwrap()` is not necessary?
C) could we return an array shape so that the `list` construct can be used? I realise this is a significant difference, however otherwise we rely on the user "knowing" that this operation will only return two elements in the collection - something that's not currently captured in the type annotations

### PR DESCRIPTION
**Pull request by drupol**
This PR

* [x] Refactor `::partition()` operation
* [x] Refactor `::span()` operation 
* [x] Has updated tests
* [x] Has updated documentation
* [x] Has static analysis tests

Depends/Follow on #125 
Fixes #123

### CODE REVIEW

**Comment by AlexandruGG** [comment]
Really nice getting this up so quickly! I will focus on the other PRs first so we can get those merged

**Comment by drupol** [comment]
I found a way to optimize the thing and remove the entries I was about to commit in `psalm-baseline.xml` !

**Comment by drupol** [comment]
Sorry for the back'n'forth, I don't touch this anymore now!

**Comment by AlexandruGG** [comment]
> Sorry for the back'n'forth, I don't touch this anymore now!

It's okay I hadn't reviewed yet 😄 . There is no change on `Span` here other than the docs, did you decide to keep it for a separate PR or does nothing need to be refactored?

**Comment by drupol** [comment]
The raw span operation is properly done, I just had to map the iterators to collection in Collection.php and it's good !

**Comment by AlexandruGG** [comment]
> The raw span operation is properly done, I just had to map the iterators to collection in Collection.php and it's good !

Sounds good, thanks for moving so fast on this. Leave this to me then I will review and add the static analysis checks

**Comment by drupol** [comment]
Sounds good :)💪👍

File: src/Contract/Operation/Spanable.php [hunk 1]
```diff
@@ -18,7 +19,9 @@
 interface Spanable
 {
     /**
-     * @return Collection<TKey, T>
+     * @param callable(T, TKey, Iterator<TKey, T>): bool $callback
```
Comment (AlexandruGG) [inline src/Contract/Operation/Spanable.php:22]: shouldn't `Span` also accept variadic callable?

**Comment by AlexandruGG** [comment]
@drupol this should be ready once the mutation tests finish (the Scrutinizer failure seems unrelated). I polished the documentation a bit, added a few more tests, and added the static analysis checks

File: src/Contract/Operation/Spanable.php (hunk 1 above)
Comment (drupol) [reply src/Contract/Operation/Spanable.php:22]: Nice catch, indeed it could be!

Comment (drupol) [reply src/Contract/Operation/Spanable.php:22]: But it could be tackled in a new PR.

**Comment by drupol** [comment]
All good, thanks !

**Comment by AlexandruGG** [comment]
> All good, thanks !

Nice, we're gonna have another big release soon 😄

**Comment by drupol** [comment]
Ho yes :)

**Pull request merged**